File: app-java/src/main/java/net/mEmoZz/yts/java/ui/detail/DetailPresenterImpl.java
package net.mEmoZz.yts.java.ui.detail;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import com.tbruyelle.rxpermissions2.RxPermissions;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;
import java.util.List;
import net.mEmoZz.yts.java.data.Quality;
import net.mEmoZz.yts.java.data.Urls;
import net.mEmoZz.yts.java.data.connection.RetroConnect;
import net.mEmoZz.yts.java.data.models.MovieDetail;
import net.mEmoZz.yts.java.data.webservice.BaseApi;
import net.mEmoZz.yts.java.ui.detail.adapters.pager.ScreenshotsAdapter;
import net.mEmoZz.yts.java.ui.detail.adapters.recycler.CastAdapter;
import net.mEmoZz.yts.java.utilities.FileUtil;
import net.mEmoZz.yts.java.utilities.Utils;
import timber.log.Timber;

/**
 * Authored by Mohamed Fathy on 08 Mar, 2018.
 * Contact: muhamed.gendy@gmail.com
 */

public class DetailPresenterImpl implements DetailPresenter {

  private DetailView detailView;
  private CompositeDisposable disposables = new CompositeDisposable();
  private MovieDetail.Movie movie;
  private List<MovieDetail.Torrent> torrents;
  private BaseApi api = new RetroConnect().initRetrofit();

  DetailPresenterImpl(DetailView detailView) {
    this.detailView = detailView;
  }

  @Override public void onAttach() {
    detailView.setActionBar();
    detailView.setupRecycler();
    detailView.initDialog();
  }

  @Override public void onDestroy() {
    disposables.dispose();
    detailView = null;
  }

  @Override public void loadMovieDetails(Activity context, long movieId) {
    if (Utils.isNetworkAvailable(context)) {
      api.getMovieDetails(movieId, true, true)
          .subscribeOn(Schedulers.io())
          .observeOn(AndroidSchedulers.mainThread())
          .doOnSubscribe(disposable -> {
            disposables.add(disposable);
            detailView.showProgressBar();
          }).doOnComplete(() -> {
        detailView.hideProgressBar();
        detailView.showMainView();
      }).subscribe(movieDetail -> {
        movie = movieDetail.getData().getMovie();
        String youtubeUrl = Urls.getYouTubeImgUrl(movie.getYoutubeCode());
        detailView.loadImages(youtubeUrl, movie.getPoster(), movie.getBackgroundImage());
        torrents = movie.getTorrents();
        setupQuality();
        setupInfo(
            movie.getYear(),
            movie.getGenres(),
            String.valueOf(movie.getRating()),
            movie.getDescription()
        );
        if (Utils.isEmpty(movie.getDescription())) detailView.hideSynopsis();
        if (isScreenshotsAvailable()) {
          detailView.setPagerAdapter(new ScreenshotsAdapter(
              movie.getScreenshotImage1(),
              movie.getScreenshotImage2(),
              movie.getScreenshotImage3()
          ));
        } else {
          detailView.hideScreenshots();
        }
        if (movie.getCast() != null && movie.getCast().size() > 0) {
          detailView.setCastRecyclerAdapter(new CastAdapter(context, movie.getCast()));
        } else {
          detailView.hideCast();
        }
      }, Timber::e);
    }
  }

  private boolean isScreenshotsAvailable() {
    return movie.getScreenshotImage1() != null
        || movie.getScreenshotImage2() != null
        || movie.getScreenshotImage3() != null;
  }

  @Override public void downloadFile(Activity context, Quality quality) {
    RxPermissions rxPermissions = new RxPermissions(context);
    rxPermissions.request(Manifest.permission.WRITE_EXTERNAL_STORAGE)
        .doOnSubscribe(disposable -> disposables.add(disposable))
        .subscribe(granted -> {
          if (granted) {
            for (int i = 0; i < torrents.size(); i++) {
              MovieDetail.Torrent torrent = torrents.get(i);
              if (torrent.getQuality().equals(quality.value)) {
                if (Utils.isNetworkAvailable(context)) {
                  api.downloadFile(torrent.getUrl())
                      .subscribeOn(Schedulers.io())
                      .observeOn(AndroidSchedulers.mainThread())
                      .doOnSubscribe(disposable -> {
                        disposables.add(disposable);
                        detailView.showProgressDialog();
                      })
                      .doOnComplete(() -> detailView.dismissProgressDialog())
                      .flatMap(FileUtil::writeResponseToDisk)
                      .subscribe(file -> detailView.showFileSavedToast(), throwable -> {
                        detailView.dismissProgressDialog();
                        detailView.showFileNotSavedToast();
                        Timber.e(throwable);
                      });
                } else {
                  detailView.showNoConnectionToast();
                }
                break;
              }
            }
          } else {
            detailView.showPermissionNotGrantedToast();
          }
        });
  }

  @Override public void copyMagnet(Context context, Quality quality) {
    for (int i = 0; i < torrents.size(); i++) {
      MovieDetail.Torrent torrent = torrents.get(i);
      if (torrent.getQuality().equals(quality.value)) {
        Utils.copyTxt(context, Urls.buildMagnetUrl(
            torrent.getHash(), movie.getTitleLong(), quality
        ));
        detailView.showCopyConfirmToast();
        break;
      }
    }
  }

  @Override public String getMovieSize(Quality quality) {
    for (int i = 0; i < torrents.size(); i++) {
      MovieDetail.Torrent torrent = torrents.get(i);
      if (torrent.getQuality().equals(quality.value)) {
        return torrent.getSize();
      }
    }
    return null;
  }

  private void setupQuality() {
    if (torrents != null && torrents.size() > 0) {
      for (int i = 0; i < torrents.size(); i++) {
        MovieDetail.Torrent torrent = torrents.get(i);
        if (torrent.is3DAvailable()) {
          detailView.enable3D();
        } else if (torrent.is720Available()) {
          detailView.enable720p();
        } else if (torrent.is1080Available()) {
          detailView.enable1080p();
        }
      }
    }
  }

  private void setupInfo(String year, List<String> genres, String rate, String description) {
    if (!Utils.isEmpty(year)) {
      detailView.showYear();
    } else {
      detailView.hideYear();
    }
    if (genres != null && genres.size() > 0) {
      detailView.showGenres();
    } else {
      detailView.hideGenres();
    }
    if (!Utils.isEmpty(rate)) {
      detailView.showRate();
    } else {
      detailView.hideRate();
    }
    detailView.setInfo(year, getGenres(genres), rate, description);
  }

  private String getGenres(List<String> genres) {
    StringBuilder builder = new StringBuilder();
    if (genres != null && genres.size() > 0) {
      for (int i = 0; i < genres.size(); i++) {
        builder.append(genres.get(i));
        if (i != genres.size() - 1) builder.append(", ");
      }
      return builder.toString();
    }
    return null;
  }
}

File: app-kotlin/src/main/java/net/mEmoZz/yts/kotlin/ui/detail/DetailPresenterImpl.kt
package net.mEmoZz.yts.kotlin.ui.detail

import android.Manifest
import android.app.Activity
import android.content.Context
import com.tbruyelle.rxpermissions2.RxPermissions
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import net.mEmoZz.yts.kotlin.data.Quality
import net.mEmoZz.yts.kotlin.data.Urls
import net.mEmoZz.yts.kotlin.data.connection.RetroConnect
import net.mEmoZz.yts.kotlin.data.models.MovieDetail
import net.mEmoZz.yts.kotlin.ui.detail.adapters.pager.ScreenshotsAdapter
import net.mEmoZz.yts.kotlin.ui.detail.adapters.recycler.CastAdapter
import net.mEmoZz.yts.kotlin.utilities.FileUtil
import net.mEmoZz.yts.kotlin.utilities.Utils
import timber.log.Timber

/**
 * Authored by Mohamed Fathy on 08 Mar, 2018.
 * Contact: muhamed.gendy@gmail.com
 */

class DetailPresenterImpl internal constructor(
    private var detailView: DetailView?
) : DetailPresenter {

  private val disposables = CompositeDisposable()
  private val api = RetroConnect().initRetrofit()

  private var movie: MovieDetail.Movie? = null
  private var torrents: List<MovieDetail.Torrent>? = null

  override fun onAttach() {
    detailView!!.setActionBar()
    detailView!!.setupRecycler()
    detailView!!.initDialog()
    detailView!!.initClicks()
  }

  override fun onDestroy() {
    disposables.dispose()
    detailView = null
  }

  override fun loadMovieDetails(context: Activity, movieId: Long) {
    if (Utils.isNetworkAvailable(context)) {
      api.getMovieDetails(movieId, true, true)
          .subscribeOn(Schedulers.io())
          .observeOn(AndroidSchedulers.mainThread())
          .doOnSubscribe({
            disposables.add(it)
            detailView!!.showProgressBar()
          }).doOnComplete({
            detailView!!.hideProgressBar()
            detailView!!.showMainView()
          }).subscribe({
            movie = it.data!!.movie
            val youtubeUrl = Urls.getYouTubeImgUrl(movie!!.youtubeCode)
            detailView!!.loadImages(youtubeUrl, movie!!.poster, movie!!.backgroundImage)
            torrents = movie!!.torrents
            setupQuality()
            setupInfo(movie!!.year, movie!!.genres, movie!!.rating.toString(), movie!!.description)
            if (Utils.isEmpty(movie!!.description)) detailView!!.hideSynopsis()
            if (isScreenshotsAvailable()) {
              detailView!!.setPagerAdapter(ScreenshotsAdapter(
                  movie!!.screenshotImage1!!,
                  movie!!.screenshotImage2!!,
                  movie!!.screenshotImage3!!
              ))
            } else {
              detailView!!.hideScreenshots()
            }
            if (movie!!.cast != null && movie!!.cast!!.isNotEmpty()) {
              detailView!!.setCastRecyclerAdapter(CastAdapter(context, movie!!.cast!!))
            } else {
              detailView!!.hideCast()
            }
          }, {
            Timber.e(it)
          })
    }
  }

  private fun isScreenshotsAvailable(): Boolean {
    return movie!!.screenshotImage1 != null
           || movie!!.screenshotImage2 != null
           || movie!!.screenshotImage3 != null
  }

  override fun downloadFile(context: Activity, quality: Quality) {
    val rxPermissions = RxPermissions(context)
    rxPermissions.request(Manifest.permission.WRITE_EXTERNAL_STORAGE)
        .doOnSubscribe { disposables.add(it) }
        .subscribe {
          if (it!!) {
            for (i in torrents!!.indices) {
              val torrent = torrents!![i]
              if (torrent.quality.equals(quality.value)) {
                if (Utils.isNetworkAvailable(context)) {
                  api.downloadFile(torrent.url!!)
                      .subscribeOn(Schedulers.io())
                      .observeOn(AndroidSchedulers.mainThread())
                      .doOnSubscribe({
                        disposables.add(it)
                        detailView!!.showProgressDialog()
                      }).doOnComplete({ detailView!!.dismissProgressDialog() })
                      .flatMap { FileUtil.writeResponseToDisk(it) }
                      .subscribe({ detailView!!.showFileSavedToast() }, {
                        detailView!!.dismissProgressDialog()
                        detailView!!.showFileNotSavedToast()
                        Timber.e(it)
                      })
                } else {
                  detailView!!.showNoConnectionToast()
                }
                break
              }
            }
          } else {
            detailView!!.showPermissionNotGrantedToast()
          }
        }
  }

  override fun copyMagnet(context: Context, quality: Quality) {
    for (i in torrents!!.indices) {
      val torrent = torrents!![i]
      if (torrent.quality.equals(quality.value)) {
        Utils.copyTxt(context, Urls.buildMagnetUrl(
            torrent.hash, movie!!.titleLong, quality
        ))
        detailView!!.showCopyConfirmToast()
        break
      }
    }
  }

  override fun getMovieSize(quality: Quality): String? {
    for (i in torrents!!.indices) {
      val torrent = torrents!![i]
      if (torrent.quality.equals(quality.value)) {
        return torrent.size
      }
    }
    return null
  }

  private fun setupQuality() {
    if (torrents != null && torrents!!.isNotEmpty()) {
      for (i in torrents!!.indices) {
        val torrent = torrents!![i]
        when {
          torrent.is3DAvailable -> detailView!!.enable3D()
          torrent.is720Available -> detailView!!.enable720p()
          torrent.is1080Available -> detailView!!.enable1080p()
        }
      }
    }
  }

  private fun setupInfo(year: String?, genres: List<String>?, rate: String?, description: String?) {
    if (!Utils.isEmpty(year)) {
      detailView!!.showYear()
    } else {
      detailView!!.hideYear()
    }
    if (genres != null && genres.isNotEmpty()) {
      detailView!!.showGenres()
    } else {
      detailView!!.hideGenres()
    }
    if (!Utils.isEmpty(rate)) {
      detailView!!.showRate()
    } else {
      detailView!!.hideRate()
    }
    detailView!!.setInfo(year, getGenres(genres), rate, description)
  }

  private fun getGenres(genres: List<String>?): String? {
    val builder = StringBuilder()
    if (genres != null && genres.isNotEmpty()) {
      for (i in genres.indices) {
        builder.append(genres[i])
        if (i != genres.size - 1) builder.append(", ")
      }
      return builder.toString()
    }
    return null
  }
}
